src/connectors/impls/gcs/writer.rs

```diff
@@ -0,0 +1,831 @@
+use async_std::channel::{bounded, Receiver, Sender};
+use async_std::task::sleep;
+#[cfg(not(test))]
+use gouth::Token;
+#[cfg(not(test))]
+use http_client::h1::H1Client;
+use http_client::HttpClient;
+use http_types::{Method, Request};
+use std::collections::HashMap;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Arc;
+use std::time::Duration;
+#[cfg(test)]
+use tests::MockHttpClient;
+use tremor_pipeline::{ConfigImpl, Event};
+use tremor_value::Value;
+use value_trait::ValueAccess;
+
+#[derive(Deserialize, Debug, Clone)]
+pub struct Config {
+    #[serde(default = "default_endpoint")]
+    endpoint: String,
+    // #[cfg_attr(test, allow(unused))]
+    #[allow(unused)] // FIXME: use it...
+    #[serde(default = "default_connect_timeout")]
+    connect_timeout: u64,
+    #[serde(default = "default_buffer_size")]
+    buffer_size: usize,
+    bucket: Option<String>,
+}
+
+fn default_endpoint() -> String {
+    "https://storage.googleapis.com/upload/storage/v1".to_string()
+}
+
+fn default_connect_timeout() -> u64 {
+    10_000_000_000
+}
+
+fn default_buffer_size() -> usize {
+    1024 * 1024 * 8 // 8MB - the recommended minimum
+}
+
+impl ConfigImpl for Config {}
+
+#[derive(Debug, Default)]
+pub(crate) struct Builder {}
+
+#[async_trait::async_trait]
+impl ConnectorBuilder for Builder {
+    fn connector_type(&self) -> ConnectorType {
+        ConnectorType("gcs_writer".into())
+    }
+
+    async fn build_cfg(
+        &self,
+        _alias: &str,
+        _config: &ConnectorConfig,
+        connector_config: &Value,
+        _kill_switch: &KillSwitch,
+    ) -> Result<Box<dyn Connector>> {
+        let config = Config::new(connector_config)?;
+
+        if config.buffer_size % (256 * 1024) != 0 {
+            return Err("Buffer size must be a multiple of 256kiB".into());
+        }
+        let url = Url::<HttpsDefaults>::parse(&config.endpoint)?;
+
+        Ok(Box::new(GCSWriterConnector { config, url }))
+    }
+}
+
+struct GCSWriterConnector {
+    config: Config,
+    url: Url<HttpsDefaults>,
+}
+
+#[async_trait::async_trait]
+impl Connector for GCSWriterConnector {
+    async fn create_sink(
+        &mut self,
+        sink_context: SinkContext,
+        builder: SinkManagerBuilder,
+    ) -> Result<Option<SinkAddr>> {
+        let default_bucket = self
+            .config
+            .bucket
+            .as_ref()
+            .map(|bucket_name| Value::from(bucket_name.as_str()).into_static());
+        let sink = GCSWriterSink {
+            client_tx: None,
+            url: self.url.clone(),
+            config: self.config.clone(),
+            buffers: ChunkedBuffer::new(self.config.buffer_size),
+            current_name: None,
+            default_bucket,
+            done_until: Arc::new(AtomicUsize::new(0)),
+        };
+
+        builder.spawn(sink, sink_context).map(Some)
+    }
+
+    fn codec_requirements(&self) -> CodecReq {
+        CodecReq::Required
+    }
+}
+
+struct ChunkedBuffer {
```

Comment:
This could need a bit of documentation on what it does and what the intent is, I've a hard time figuring that out just from the code